articles/remapping.md

```diff
@@ -0,0 +1,234 @@
+- *TODO Global topics like /tf and /tf_static must be individually remapped in ROS 1. What to do about them?*
+
+
+#### A user can remap a name they see on a node after it has launched
+This implies remapping rules are applied in sequence.
+This also implies there is a way to supply remapping rules to a node besides the command line (services?).
+
+**Rationale:**
+A developer may want to make a running node subscribe to a different source of data to see how it behaves.
+It is less work for them to change the name they see rather than digging through the source code to find the original name.
+
+
+### Structure of a Remapping Rule
+Remapping rules have two parts:
+
+1. a part used to determine if the rule applies to a name
+2. a replacement for the names that are matched
+
+
+### Remapping Names in ROS 1
+Remapping is a feature that also exists in ROS 1.
+In ROS 1 remapping works by passing in [arguments](http://wiki.ros.org/Remapping%20Arguments) to each node.
+Client implementations also have APIs in code to pass remapping rules when the node is initialized.
+A remap rule consists of two names: one that should be replaced with another.
+
+ROS 1 remapping works on **Fully Qualified Names** (FQN).
+Both sides of a rule are [expanded to FQN](http://docs.ros.org/api/roscpp/html/namespaceros_1_1names.html#a377ff8fede7b95398fd5d1c5cd49246b).
+Before a name is remapped it is also [expanded to FQN](http://docs.ros.org/api/roscpp/html/namespaceros_1_1names.html#ab2eebaf734abfbdccb4122f8e24f547f).
+The name is remapped to the right side only if it exactly matches the left side of a rule.
+
+
+### Thoughts on ROS 2 implementation
+`rmw` is likely ignorant of remapping.
+The APIs offered by client implementations should remap names automatically to prevent creation of names that are impossible to remap.
+
+
+### Remapping rule command line syntax
+In ROS 1 remapping is done primarily through command line arguments, commonly via roslaunch.
+It may not be possible to use the same syntax for ROS 2 because multiple nodes can be inside of the same process, and it is desired that remap rules are node specific and not process specific.
+This syntax assumes the node's name is guaranteed to be unique in a process.
+It does not support pre-FQN expansion remapping.
+
+- `%`
+ - During matching it is a wildcard that matches a single token
+ - It may appear at most once on the matching side of a rule
+ - During replacement it is replaced with the matched token
+
+- `%%`
+ - During matching it is a wildcard that matches zero or more tokens with at least one slash
+ - It may appear at most once on the matching side of a rule
+ - During replacement it is replaced with the matched slashes and tokens
+
+- `:=` divides the two parts of a remapping rule: matching and replacement
+
+- `~` expands to `/namespace/nodename`
+
+- `nodename|` prefixed to a rule makes it apply only to a node with that unique name
+ - If not specified, the rule applies to all nodes in the process
```

Comment:
Again, a rationale for the syntax selection would be nice to see here. I don't have a better suggestion off-hand, but it would be nice to see this justified (if there is one). Also will using `|` make it hard to type this in a shell?

I think this is a fine way to narrow down individual rules to individual nodes. However, I was thinking about this a lot this weekend and I was unable to convince myself that we "must" provide this feature. The only time you would need to provide different remapping arguments to a single executable with multiple nodes via the command line would be when someone has written their own main function and instantiated their own nodes. This would be a nice feature to provide these people since they would have to implement the argument parsing on their own otherwise, but I'm not convinced it will be a commonplace thing.

I think it's far more likely that two other modes of operation will be used frequently, and neither would require passing remapping for multiple nodes over the command line.

The first mode is a single node in a single executable. I expect this to be commonplace, but the ROS 1 style command line arguments should work here.

The second mode will be that a "container" is run for multiple nodes (same thing as Nodelet manager in ROS 1) and then the nodes it runs will be loaded via a configuration file or after it has started using ROS services. The process would be: start a container, add some stuff to it, and specifying the remappings for each as you add them.

So I think we should go through the exercise of coming up with a syntax, but it doesn't have to be especially fancy in my opinion because I believe it should be rarely used. That being said, I think the proposed syntax should work out fine.